File: agenkit-ts/examples/adapters/litellm-basic.ts
/**
 * Basic LiteLLM example
 *
 * Demonstrates:
 * - LiteLLM adapter configuration
 * - Single-turn completion via LiteLLM proxy
 * - Multi-turn conversation
 * - Streaming responses
 * - Multiple provider support
 *
 * Setup:
 *   1. Install and run LiteLLM proxy:
 *      pip install litellm
 *      litellm --model gpt-3.5-turbo
 *   2. Or use Docker:
 *      docker run -p 4000:4000 ghcr.io/berriai/litellm:latest
 *   3. Run this example:
 *      npm run build
 *      node dist/examples/litellm-basic.js
 */

import { LiteLLMAdapter } from '../src/adapters/litellm';
import { createMessage } from '../src/core/interfaces';

async function main() {
  console.log('='.repeat(60));
  console.log('AgentKit TypeScript - LiteLLM Example');
  console.log('='.repeat(60));
  console.log();

  console.log('Make sure LiteLLM proxy is running:');
  console.log('  litellm --model gpt-3.5-turbo');
  console.log('  or');
  console.log('  docker run -p 4000:4000 ghcr.io/berriai/litellm:latest');
  console.log();

  // Create LiteLLM adapter
  const adapter = new LiteLLMAdapter({
    baseUrl: 'http://localhost:4000',
    model: 'gpt-3.5-turbo',
    temperature: 0.7,
    maxTokens: 1024,
  });

  console.log(`✓ Initialized ${adapter.name}`);
  console.log(`✓ Capabilities: ${adapter.capabilities.join(', ')}`);
  console.log(`✓ Proxy: http://localhost:4000`);
  console.log();

  // Example 1: Simple completion
  console.log('-'.repeat(60));
  console.log('Example 1: Simple Completion');
  console.log('-'.repeat(60));
  console.log();

  const message1 = createMessage({
    role: 'user',
    content: 'Explain the concept of a universal LLM gateway in one paragraph.',
  });

  console.log(`User: ${message1.content}`);
  console.log();

  try {
    const response1 = await adapter.process(message1);
    console.log(`LiteLLM: ${response1.content}`);
    console.log();
    console.log('Metadata:');
    console.log(`  Model: ${response1.metadata?.model}`);
    console.log(`  Tokens: ${JSON.stringify(response1.metadata?.usage)}`);
    console.log(`  Finish reason: ${response1.metadata?.finish_reason}`);
    console.log();
  } catch (error) {
    console.error(`Error: ${error instanceof Error ? error.message : 'Unknown error'}`);
    console.error('Make sure LiteLLM proxy is running on http://localhost:4000');
    process.exit(1);
  }

  // Example 2: With system message
  console.log('-'.repeat(60));
  console.log('Example 2: Conversation with System Message');
  console.log('-'.repeat(60));
  console.log();

  const conversation = [
    createMessage({
      role: 'system',
      content: 'You are a helpful assistant that explains technical concepts clearly.',
    }),
    createMessage({
      role: 'user',
      content: 'What is LiteLLM?',
    }),
  ];

  console.log('System: You are a helpful assistant that explains technical concepts clearly.');
  console.log('User: What is LiteLLM?');
  console.log();

  const response2 = await adapter.complete(conversation);
  console.log(`LiteLLM: ${response2.content}`);
  console.log();
  console.log(`Input tokens: ${response2.metadata?.usage?.prompt_tokens}`);
  console.log(`Output tokens: ${response2.metadata?.usage?.completion_tokens}`);
  console.log();

  // Example 3: Streaming response
  console.log('-'.repeat(60));
  console.log('Example 3: Streaming Response');
  console.log('-'.repeat(60));
  console.log();

  const message3 = createMessage({
    role: 'user',
    content: 'List the top 5 benefits of using a universal LLM gateway.',
  });

  console.log(`User: ${message3.content}`);
  console.log();
  console.log('LiteLLM (streaming): ');

  let fullResponse = '';
  for await (const chunk of adapter.processStream(message3)) {
    process.stdout.write(chunk.content as string);
    fullResponse += chunk.content;
  }

  console.log();
  console.log();
  console.log(`Full response length: ${fullResponse.length} characters`);
  console.log();

  // Example 4: Different models through LiteLLM
  console.log('-'.repeat(60));
  console.log('Example 4: Using Different Providers');
  console.log('-'.repeat(60));
  console.log();

  console.log('LiteLLM supports 100+ providers:');
  console.log('  OpenAI: gpt-4, gpt-3.5-turbo');
  console.log('  Anthropic: claude-3-5-sonnet-20241022');
  console.log('  AWS Bedrock: bedrock/anthropic.claude-v2');
  console.log('  Google Gemini: gemini/gemini-pro');
  console.log('  Azure OpenAI: azure/gpt-4');
  console.log('  Local Ollama: ollama/llama2');
  console.log();

  // Example with GPT-4 (if available)
  console.log('Switching to GPT-4 (if configured in LiteLLM):');
  const gpt4Adapter = new LiteLLMAdapter({
    baseUrl: 'http://localhost:4000',
    model: 'gpt-4',
    temperature: 0.7,
    maxTokens: 200,
  });

  const message4 = createMessage({
    role: 'user',
    content: 'What are microservices in one sentence?',
  });

  console.log(`User: ${message4.content}`);
  console.log();

  try {
    const response4 = await gpt4Adapter.process(message4);
    console.log(`GPT-4 (via LiteLLM): ${response4.content}`);
    console.log();
  } catch (error) {
    console.log('Note: GPT-4 model not configured in LiteLLM proxy');
    console.log();
  }

  // Example 5: Multi-turn dialogue
  console.log('-'.repeat(60));
  console.log('Example 5: Multi-Turn Dialogue');
  console.log('-'.repeat(60));
  console.log();

  const dialogue = [
    createMessage({ role: 'user', content: 'What is API rate limiting?' }),
  ];

  console.log('User: What is API rate limiting?');

  const turn1 = await adapter.complete(dialogue);
  console.log(`LiteLLM: ${turn1.content}`);
  console.log();

  dialogue.push(createMessage({ role: 'assistant', content: turn1.content as string }));
  dialogue.push(createMessage({ role: 'user', content: 'How does LiteLLM help with rate limiting?' }));

  console.log('User: How does LiteLLM help with rate limiting?');

  const turn2 = await adapter.complete(dialogue);
  console.log(`LiteLLM: ${turn2.content}`);
  console.log();

  console.log(`Total tokens in conversation: ${turn2.metadata?.usage?.total_tokens}`);
  console.log();

  // Example 6: With authentication
  console.log('-'.repeat(60));
  console.log('Example 6: Authenticated LiteLLM Proxy');
  console.log('-'.repeat(60));
  console.log();

  console.log('For production deployments, use API key authentication:');
  console.log();
  console.log('const secureAdapter = new LiteLLMAdapter({');
  console.log('  baseUrl: "https://your-litellm-proxy.com",');
  console.log('  model: "gpt-4",');
  console.log('  apiKey: process.env.LITELLM_API_KEY,');
  console.log('});');
  console.log();

  // Example 7: Error handling
  console.log('-'.repeat(60));
  console.log('Example 7: Error Handling');
  console.log('-'.repeat(60));
  console.log();

  const invalidAdapter = new LiteLLMAdapter({
    baseUrl: 'http://localhost:4000',
    model: 'non-existent-model',
  });

  try {
    await invalidAdapter.process(createMessage({
      role: 'user',
      content: 'Test',
    }));
  } catch (error) {
    console.log('Caught expected error:');
    console.log(`  ${error instanceof Error ? error.message : 'Unknown error'}`);
    console.log();
    console.log('LiteLLM adapter properly handles:');
    console.log('  - Invalid model names');
    console.log('  - Network errors');
    console.log('  - Timeout errors');
    console.log('  - API authentication errors');
    console.log();
  }

  console.log('-'.repeat(60));
  console.log('✓ All examples completed successfully!');
  console.log('-'.repeat(60));
  console.log();
  console.log('Next steps:');
  console.log('  1. Configure LiteLLM with your preferred providers');
  console.log('  2. Set up load balancing across multiple models');
  console.log('  3. Enable caching for cost optimization');
  console.log('  4. Monitor usage with LiteLLM dashboard');
}

main().catch((error) => {
  console.error('Error:', error.message);
  console.error('');
  console.error('Make sure LiteLLM proxy is running:');
  console.error('  pip install litellm');
  console.error('  litellm --model gpt-3.5-turbo');
  console.error('');
  process.exit(1);
});

File: agenkit-cpp/examples/adapters/litellm-basic.cpp
/**
 * @file litellm-basic.cpp
 * @brief Basic LiteLLM proxy usage example
 *
 * Demonstrates:
 * - LiteLLM adapter configuration
 * - Universal gateway access to multiple providers
 * - Single-turn completion
 * - Error handling
 * - Metadata extraction
 *
 * Setup:
 *   # Start LiteLLM proxy
 *   litellm --port 4000
 *
 *   # Or with specific providers
 *   export OPENAI_API_KEY=your-key
 *   litellm --model gpt-3.5-turbo --port 4000
 *
 *   # Build and run
 *   cmake -B build -S .
 *   cmake --build build
 *   ./build/examples/litellm-basic
 */

#include "agenkit/adapters/litellm_agent.hpp"
#include "agenkit/core/message.hpp"
#include <iostream>
#include <cstdlib>

using namespace agenkit;
using namespace agenkit::adapters;
using namespace agenkit::core;

void print_separator(const std::string& title = "") {
    std::cout << "\n";
    std::cout << std::string(60, '=') << "\n";
    if (!title.empty()) {
        std::cout << title << "\n";
        std::cout << std::string(60, '=') << "\n";
    }
    std::cout << "\n";
}

int main() {
    print_separator("AgentKit C++ - LiteLLM Basic Example");

    std::cout << "LiteLLM is a universal LLM gateway supporting 100+ providers:\n";
    std::cout << "  • OpenAI (gpt-4, gpt-3.5-turbo)\n";
    std::cout << "  • Anthropic (claude-3-5-sonnet-20241022)\n";
    std::cout << "  • AWS Bedrock (bedrock/anthropic.claude-v2)\n";
    std::cout << "  • Google Gemini (gemini/gemini-pro)\n";
    std::cout << "  • Azure OpenAI (azure/gpt-4)\n";
    std::cout << "  • Cohere (command-r-plus)\n";
    std::cout << "  • Local models (ollama/llama2, ollama/mistral)\n";
    std::cout << "  • And 100+ more!\n\n";

    std::cout << "✓ Initialized LiteLLM adapter\n\n";

    // Example 1: Simple completion
    print_separator("Example 1: Simple Completion");

    try {
        LiteLLMConfig config;
        config.base_url = "http://localhost:4000";
        config.model = "gpt-3.5-turbo";
        config.temperature = 0.7;
        config.max_tokens = 1024;

        // Check for optional API key
        const char* api_key = std::getenv("LITELLM_API_KEY");
        if (api_key) {
            config.api_key = api_key;
            std::cout << "✓ Using API key from LITELLM_API_KEY\n";
        }

        LiteLLMAgent agent(config);

        std::cout << "Model: " << config.model << "\n";
        std::cout << "Base URL: " << config.base_url << "\n";
        std::cout << "User: Explain what LiteLLM is in two sentences.\n\n";

        auto msg = Message::with_text("user", "Explain what LiteLLM is in two sentences.");
        auto future = agent.process(std::move(msg));
        auto result = future.get();

        if (result.is_ok()) {
            const auto& response = result.unwrap();
            std::cout << "Assistant: " << response.content_as_str() << "\n\n";

            // Show metadata
            std::cout << "Metadata:\n";
            const auto& meta = response.metadata();
            if (meta.contains("model")) {
                std::cout << "  Model: " << meta["model"].get<std::string>() << "\n";
            }
            if (meta.contains("finish_reason")) {
                std::cout << "  Finish reason: " << meta["finish_reason"].get<std::string>() << "\n";
            }
            if (meta.contains("usage")) {
                const auto& usage = meta["usage"];
                if (usage.contains("total_tokens")) {
                    std::cout << "  Total tokens: " << usage["total_tokens"].get<int>() << "\n";
                }
            }
        } else {
            std::cerr << "Error: " << result.unwrap_err().message() << "\n";
        }
    } catch (const std::exception& e) {
        std::cerr << "Exception: " << e.what() << "\n";
    }

    // Example 2: Multiple provider routing
    print_separator("Example 2: Multiple Provider Routing");

    try {
        std::string prompt = "What is the meaning of life?";

        // Each model can be routed to different providers by LiteLLM
        std::vector<std::string> models = {
            "gpt-3.5-turbo",           // OpenAI
            "claude-3-haiku-20240307", // Anthropic
            "gemini/gemini-pro"        // Google Gemini
        };

        for (const auto& model : models) {
            LiteLLMConfig config;
            config.base_url = "http://localhost:4000";
            config.model = model;
            config.max_tokens = 100;

            LiteLLMAgent agent(config);

            std::cout << "Model: " << model << "\n";
            std::cout << "User: " << prompt << "\n\n";

            auto msg = Message::with_text("user", prompt);
            auto future = agent.process(std::move(msg));
            auto result = future.get();

            if (result.is_ok()) {
                std::cout << "Response: " << result.unwrap().content_as_str() << "\n\n";
            } else {
                std::cout << "Error: " << result.unwrap_err().message() << "\n\n";
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "Exception: " << e.what() << "\n";
    }

    // Example 3: Temperature variation
    print_separator("Example 3: Temperature Effects");

    try {
        std::string prompt = "Write a haiku about programming.";

        for (double temp : {0.3, 0.7, 1.2}) {
            LiteLLMConfig config;
            config.base_url = "http://localhost:4000";
            config.model = "gpt-3.5-turbo";
            config.temperature = temp;
            config.max_tokens = 50;

            LiteLLMAgent agent(config);

            std::cout << "Temperature: " << temp << "\n";
            std::cout << "User: " << prompt << "\n\n";

            auto msg = Message::with_text("user", prompt);
            auto future = agent.process(std::move(msg));
            auto result = future.get();

            if (result.is_ok()) {
                std::cout << "Response:\n" << result.unwrap().content_as_str() << "\n\n";
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "Exception: " << e.what() << "\n";
    }

    // Example 4: Using model constants
    print_separator("Example 4: Model Constants");

    try {
        std::cout << "Available model constants:\n\n";

        LiteLLMConfig config;
        config.base_url = "http://localhost:4000";
        config.model = LiteLLMModels::GPT_4O_MINI;
        config.max_tokens = 50;

        LiteLLMAgent agent(config);

        std::cout << "Using: " << LiteLLMModels::GPT_4O_MINI << "\n";
        std::cout << "User: Hello, how are you?\n\n";

        auto msg = Message::with_text("user", "Hello, how are you?");
        auto future = agent.process(std::move(msg));
        auto result = future.get();

        if (result.is_ok()) {
            std::cout << "Response: " << result.unwrap().content_as_str() << "\n\n";
        }

        std::cout << "Other available constants:\n";
        std::cout << "  OpenAI: " << LiteLLMModels::GPT_4 << ", "
                  << LiteLLMModels::GPT_4_TURBO << "\n";
        std::cout << "  Anthropic: " << LiteLLMModels::CLAUDE_3_5_SONNET << "\n";
        std::cout << "  Bedrock: " << LiteLLMModels::BEDROCK_CLAUDE_V2 << "\n";
        std::cout << "  Gemini: " << LiteLLMModels::GEMINI_PRO << "\n";
        std::cout << "  Ollama: " << LiteLLMModels::OLLAMA_LLAMA2 << "\n";
    } catch (const std::exception& e) {
        std::cerr << "Exception: " << e.what() << "\n";
    }

    // Example 5: Error handling
    print_separator("Example 5: Error Handling");

    try {
        // Test with unreachable server
        LiteLLMConfig config;
        config.base_url = "http://localhost:9999";  // Wrong port
        config.model = "gpt-3.5-turbo";
        config.timeout = std::chrono::milliseconds{2000};

        LiteLLMAgent agent(config);

        std::cout << "Testing with unreachable server (localhost:9999)...\n\n";

        auto msg = Message::with_text("user", "Hello");
        auto future = agent.process(std::move(msg));
        auto result = future.get();

        if (result.is_err()) {
            std::cout << "✓ Error caught as expected:\n";
            std::cout << "  Type: " << static_cast<int>(result.unwrap_err().type()) << "\n";
            std::cout << "  Message: " << result.unwrap_err().message() << "\n";
        } else {
            std::cout << "⚠️  Unexpected success with unreachable server\n";
        }
    } catch (const std::exception& e) {
        std::cout << "✓ Exception caught: " << e.what() << "\n";
    }

    print_separator("✓ All examples completed!");

    std::cout << "Key Features Demonstrated:\n";
    std::cout << "  • LiteLLM adapter configuration\n";
    std::cout << "  • Universal gateway for multiple providers\n";
    std::cout << "  • Single-turn completion\n";
    std::cout << "  • Temperature effects\n";
    std::cout << "  • Model constants\n";
    std::cout << "  • Error handling and metadata\n\n";

    std::cout << "Next Steps:\n";
    std::cout << "  • Try different models by changing config.model\n";
    std::cout << "  • Experiment with temperature and max_tokens\n";
    std::cout << "  • Use LiteLLM to route to AWS Bedrock, Azure, etc.\n";
    std::cout << "  • See https://docs.litellm.ai/ for more info\n\n";

    return 0;
}
